### PR TITLE
History viewed cars implementation

### PR DESCRIPTION
**Pull request by Aleksandr-Shulha-dev**

### CODE REVIEW

File: backend/src/controllers/reviewed-cars/reviewed-cars.controller.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    const { userId } = req.params;
+    const { modelId, complectationId } = req.query;
+
+    const reviewedCarsList = {
+      userId,
+      modelId,
+      complectationId,
+    };
+
+    const result = await reviewedCarsService.setCar(reviewedCarsList);
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/reviewed-cars/reviewed-cars.controller.ts]: `reviewed` - машини про котрі залишили відгук
`viewed` - проглянуті

File: backend/src/services/reviewed-cars/reviewed-cars.service.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    select: {
+      model_id: true,
+      complectation_id: true,
+    },
+  });
+
+  const detailedListOfReviewedCars = await Promise.all(
+    viewedСarsList.map(async ({ model_id, complectation_id }) => {
+      if (model_id && complectation_id) {
+        return await prisma.model.findFirst({
```
Comment (oleksandr-danylchenko) [inline backend/src/services/reviewed-cars/reviewed-cars.service.ts]: `return await` is redundant. Only `return` can be used, the calling side will await

File: backend/src/services/reviewed-cars/reviewed-cars.service.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    },
+    select: {
+      model_id: true,
+      complectation_id: true,
+    },
+  });
+
+  const detailedListOfReviewedCars = await Promise.all(
+    viewedСarsList.map(async ({ model_id, complectation_id }) => {
+      if (model_id && complectation_id) {
```
Comment (oleksandr-danylchenko) [inline backend/src/services/reviewed-cars/reviewed-cars.service.ts]: It should be impossible for the `model_id` and the `complectation_id` be not null at the same time in the `users_Viewed_Cars` table

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]
Some wording update is required + some left some questions

File: backend/src/controllers/reviewed-cars/reviewed-cars.controller.ts (hunk 1 above)
Comment (oleksandr-shulha-dev) [reply backend/src/controllers/reviewed-cars/reviewed-cars.controller.ts]: I renamed the service and the controller on the viewed cars

File: backend/src/services/reviewed-cars/reviewed-cars.service.ts (hunk 2 above)
Comment (oleksandr-shulha-dev) [reply backend/src/services/reviewed-cars/reviewed-cars.service.ts]: Done

File: backend/src/services/reviewed-cars/reviewed-cars.service.ts (hunk 3 above)
Comment (oleksandr-shulha-dev) [reply backend/src/services/reviewed-cars/reviewed-cars.service.ts]: I fixed that

File: backend/src/controllers/viewed-cars/viewed-cars.controller.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import type {
+  setViewedCarRequest,
+  setViewedCarResponse,
+} from '@autoline/shared';
+import type { TypedRequestQuery } from '@common/types/controller/controller';
+import type { getViewedCarsListResponse } from '@common/types/types';
+import type { WishlistInput } from '@common/types/types';
+import type { NextFunction, Response } from 'express';
+
+const getList = async (
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/viewed-cars/viewed-cars.controller.ts]: Please use more explicit naming. E.g. `getViewedCarsList`

File: backend/src/controllers/viewed-cars/viewed-cars.controller.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  } catch (error) {
+    console.error(error);
+    if (error instanceof Error) {
+      res.sendStatus(400);
+    }
+    next(error);
+  }
+};
+
+const setCar = async (
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/viewed-cars/viewed-cars.controller.ts]: Please use more explicit naming. E.g. `addCarToViewed`

File: backend/src/controllers/viewed-cars/viewed-cars.controller.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+): Promise<void> => {
+  try {
+    const { userId } = req.params;
+    const { modelId, complectationId } = req.query;
+
+    const reviewedCarsList = {
+      userId,
+      modelId,
+      complectationId,
+    };
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/viewed-cars/viewed-cars.controller.ts]: This object is misnamed. It's not the `List`

File: backend/src/services/viewed-cars/viewed-cars.service.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
+import { prisma } from '@data/prisma-client';
+
+import type {
+  setViewedCarRequest,
+  setViewedCarResponse,
+} from '@autoline/shared';
+import type { getViewedCarsListResponse } from '@common/types/types';
+
+const getList = async (
```
Comment (oleksandr-danylchenko) [inline backend/src/services/viewed-cars/viewed-cars.service.ts]: Explicitness

File: backend/src/services/viewed-cars/viewed-cars.service.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    orderBy: {
+      created_at: 'desc',
+    },
+    select: {
+      model_id: true,
+      complectation_id: true,
+    },
+  });
+
+  const detailedListOfReviewedCars = await Promise.all(
```
Comment (oleksandr-danylchenko) [inline backend/src/services/viewed-cars/viewed-cars.service.ts]: `reviewed` -> `viewed`

File: backend/src/services/viewed-cars/viewed-cars.service.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+          },
+        },
+      });
+    }),
+  );
+
+  return detailedListOfReviewedCars;
+};
+
+const setCar = async ({
```
Comment (oleksandr-danylchenko) [inline backend/src/services/viewed-cars/viewed-cars.service.ts]: Explicitness

File: backend/src/services/viewed-cars/viewed-cars.service.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+}: setViewedCarRequest): Promise<setViewedCarResponse> => {
+  const reviewedCarList = await prisma.users_Viewed_Cars.findFirst({
+    where: {
+      user_id: userId,
+      model_id: modelId,
+      complectation_id: complectationId,
+    },
+  });
+
+  if (reviewedCarList) {
```
Comment (oleksandr-danylchenko) [inline backend/src/services/viewed-cars/viewed-cars.service.ts]: `reviewed` -> `viewed`

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]

File: backend/src/controllers/viewed-cars/viewed-cars.controller.ts (hunk 4 above)
Comment (oleksandr-shulha-dev) [reply backend/src/controllers/viewed-cars/viewed-cars.controller.ts]: done

File: backend/src/controllers/viewed-cars/viewed-cars.controller.ts (hunk 5 above)
Comment (oleksandr-shulha-dev) [reply backend/src/controllers/viewed-cars/viewed-cars.controller.ts]: done

File: backend/src/controllers/viewed-cars/viewed-cars.controller.ts (hunk 6 above)
Comment (oleksandr-shulha-dev) [reply backend/src/controllers/viewed-cars/viewed-cars.controller.ts]: done

File: backend/src/services/viewed-cars/viewed-cars.service.ts (hunk 7 above)
Comment (oleksandr-shulha-dev) [reply backend/src/services/viewed-cars/viewed-cars.service.ts]: done

File: backend/src/services/viewed-cars/viewed-cars.service.ts (hunk 8 above)
Comment (oleksandr-shulha-dev) [reply backend/src/services/viewed-cars/viewed-cars.service.ts]: done

File: backend/src/services/viewed-cars/viewed-cars.service.ts (hunk 9 above)
Comment (oleksandr-shulha-dev) [reply backend/src/services/viewed-cars/viewed-cars.service.ts]: done

**Comment by oleksandr-danylchenko** [comment]
Please resolve discussions instead of just leaving answers. It'll make my notifications inbox slimmer

Thanks :slightly_smiling_face:

**Review by oleksandr-danylchenko — APPROVED** [review]

**Pull request merged**